Use asyncio in XiaomiAqara instead of threading

starkillerOG opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Use async instead of a infinite while loop in a thread for listening to multicast messages in the Xiaomi Aqara integration.
Bump PyXiaomiGateway to 0.14.1:
https://github.com/Danielhiversen/PyXiaomiGateway/compare/0.13.4...0.14.1

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #https://github.com/home-assistant/core/issues/70437 https://github.com/home-assistant/core/issues/69056
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

starkillerOG committed:
use async instead of threading

starkillerOG committed:
fix styling

starkillerOG committed:
fix gateway discovery

starkillerOG commented:
Test failure is unrelated

starkillerOG commented:
@Danielhiversen would you mind revieuwing and approving this PR?

starkillerOG commented:
@frenck do you have time to review/merge this?

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/__init__.py:
```diff
@@ -158,24 +162,27 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+            def stop_xiaomi(event):
+                """Stop Xiaomi Socket."""
+                _LOGGER.debug("Shutting down Xiaomi Gateway Listener")
+                multicast.stop_listen()
```
Why is this run in the thread pool?

starkillerOG replied:
I changed it to async thanks!

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/__init__.py:
```diff
@@ -201,23 +208,32 @@ def stop_xiaomi(event):
-        gateway_discovery = hass.data[DOMAIN].pop(LISTENER_KEY)
-        await hass.async_add_executor_job(gateway_discovery.stop_listen)
+        multicast = hass.data[DOMAIN].pop(LISTENER_KEY)
+        await hass.async_add_executor_job(multicast.stop_listen)
```
Why is this run in the thread pool?

starkillerOG replied:
Because I did not know if `self.transport.close()` and `self._sock.close()` were async.
But I now know that they can and schould be run from within the event loop.
Thanks!

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/manifest.json:
```diff
@@ -3,7 +3,7 @@
   "config_flow": true,
   "documentation": "https://www.home-assistant.io/integrations/xiaomi_aqara",
-  "requirements": ["PyXiaomiGateway==0.13.4"],
+  "requirements": ["PyXiaomiGateway==0.14.1"],
```
The library should close the transport in the event loop.

`loop.create_datagram_endpoint` returns a tuple with the transport and protocol. The library is naming the variables in a confusing way.

https://github.com/Danielhiversen/PyXiaomiGateway/blob/d0214c2ba52a9623b898e8c80ebc8d326b830775/xiaomi_gateway/__init__.py#L122-L123

It's not a coroutine. It's the protocol. And why not await `_create_udp_listener` directly?

starkillerOG replied:
Thank you very much for explaining, I am learning day by day!

I think I am now indeed closing the transport in the event loop.
I made a upstream library PR to clearify the naming here: https://github.com/Danielhiversen/PyXiaomiGateway/pull/197 (but we do not have to wait for that since it does not change the functioning).

That was because I used examples from other libraries that I addepted, I am still learning a lot about async.

Thank you for your help!

starkillerOG committed:
stop_xiaomi async

starkillerOG committed:
Stop listen async

starkillerOG committed:
Do not use add_job

starkillerOG commented:
@MartinHjelmare I prossed all your feedback, thanks a lot, I keep learing.
I also removed the add_job and changed bus.fire() to bus.async_fire() since those are all callbacks from the datagram_protocol so within the event loop (I think).

Do you have any more feedback or can this be merged?

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/binary_sensor.py:
```diff
@@ -293,7 +293,7 @@ def parse_data(self, data, raw_data):
 
             if self.entity_id is not None:
-                self._hass.bus.fire(
+                self._hass.bus.async_fire(
```
Side note: All of these events should be moved out of the binary sensor platform. Entities shouldn't fire custom events.

starkillerOG replied:
I know (from the Xiaomi Miio PR) but that is out of scope for this PR.
The rest of that part of the code is untouched, I only use the appropriate async_fire instead of just fire.

MartinHjelmare commented on homeassistant/components/xiaomi_aqara/__init__.py:
```diff
@@ -158,24 +162,27 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
-            _LOGGER.debug("Shutting down Xiaomi Gateway Listener")
-            gateway_discovery.stop_listen()
+            # register stop callback to shutdown listining for local pushes
+            async def stop_xiaomi(event):
```
```suggestion
            @callback
            def stop_xiaomi(event):
```

starkillerOG committed:
Update homeassistant/components/xiaomi_aqara/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

starkillerOG commented:
@MartinHjelmare can you merge this?

MartinHjelmare reviewed (approved):
Thanks!
